working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+`final` may also occur as a modifier on the class itself, and also because
+the resulting class header does not contain syntax which is already similar
+to a non-primary constructor declaration.
+`class final Point(int x, int y);` cannot use the similarity to a
+non-primary constructor declaration to justify the keyword `final`.
+
+```dart
+class Point {
+  final int x;
+  final int y;
+  Point(this.x, this.y);
+}
+
+class final Point(int x, int y); // Not supported!
+```
+
+Optional parameters can be declared as usual in a primary constructor, with
+default values that must be constant as usual:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, [this.y = 0]);
+}
+
+class Point(int x, [int y = 0]);
+```
+
+Similarly for named parameters, required or not:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, {required this.y});
+}
+
+class Point(int x, {required int y});
+```
+
+The class header can have additional elements, just like class headers
+where there is no primary constructor:
+
+```dart
+class D<TypeVariable extends Bound> extends A with M implements B, C {
+  final int x;
+  final int y;
+  const D.named(this.x, [this.y = 0]);
+}
+
+class const D.named<TypeVariable extends Bound>(int x, [int y = 0])
+    extends A with M implements B, C;
+```
+
+## Specification
+
+### Syntax
+
+The grammar is modified as follows. Note that the changes include
+support for inline classes, because they're intended to use primary
+constructors as well.
+
+```
+ <topLevelDefinition> ::=
+     <classDeclaration>
+   | <inlineClassDeclaration> // New alternative.
+   | ...;
+
+<classDeclaration> ::= // First alternative modified.
+     (<classModifiers> | <mixinClassModifiers>)
+     'class' <classNamePart> <superclass>? <interfaces>? <classBody>
+   | ...;
+
+<classNamePart> ::= // New rule.
+     'const'? <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+
+<classBody> ::= // New rule.
+     '{' (<metadata> <classMemberDeclaration>)* '}'
+   | ';';
+
+<inlineClassDeclaration> ::=
+     'final'? 'inline' 'class' <classNamePart> <interfaces>? <inlineClassBody>;
+
+<inlineClassBody> ::=
+     '{' (<metadata> <inlineMemberDeclaration>)* '}'
+   | ';';
+
+<inlineMemberDeclaration> ::= <classMemberDeclaration>;
+
+<enumType> ::= // Modified rule.
+     'enum' <enumNamePart> <mixins>? <interfaces>? '{'
+        <enumEntry> (',' <enumEntry>)* (',')?
+        (';' (<metadata> <classMemberDeclaration>)*)?
+     '}';
+
+<enumNamePart> ::= // New rule.
+     <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+```
+
+The word `inline` is now used in the grammar, but it is not a reserved word
+or a built-in identifier.
+
+### Static processing
+
+A class declaration with a primary constructor *(including `inline`, but
+not `<mixinApplicationClass>`, because the grammar rules do not include
+those)* is desugared to a class declaration without a primary
```

Comment:
> I really, really want to avoid "desugaring" if possible.

Of course, we can just outlaw that approach in general. However, I don't think it's helpful in this case.

I actually think we want to be able to promise that a simple rewrite operation will show exactly how a primary constructor works. If we insist on using natural language and no desugaring then we need to start entirely from scratch describing how a primary constructor works, and we're going to create the impression that this is a new primitive that provides support for things that we could not express previously. That's just misleading.

The danger associated with desugaring is that it specifies many properties simultaneously. We usually cut down on this by specifying the static analysis of the pre-desugaring syntax separately, and we must then be careful to ensure that the desugaring step transforms every program with no compile-time errors into a program with no compile-time errors. This makes it possible to rely on the dynamic semantics of the desugared program, and it also supports the property that the static analysis can ignore the desugaring step.

> We'll need to say _when_ the rewrite occurs. There is a program before rewrite, and a program after rewrite.

True. I believe the approach that I've outlined above is similar to the approach that we're taking in similar situations, and it has worked so far. For example, shorting `?.` chains is a much, much more complex mechanism, but we haven't specified that in terms of a direct semantics, we've used a rewriting.

Desugaring is a slippery rope in general, but I think we should use it judiciously rather than strictly avoiding it, and I don't think this feature is among the most tricky ones in that respect.